zrml/market-commons/src/migrations.rs

```diff
@@ -14,3 +14,281 @@
+    pub creator: AI,
+    pub creation: MarketCreation,
+    pub creator_fee: u8,
+    pub oracle: AI,
+    pub metadata: Vec<u8>,
+    pub market_type: MarketType,
+    pub period: MarketPeriod<BN, M>,
+    pub scoring_rule: ScoringRule,
+    pub status: MarketStatus,
+    pub report: Option<Report<AI, BN>>,
+    pub resolved_outcome: Option<OutcomeReport>,
+    pub dispute_mechanism: LegacyMarketDisputeMechanism<AI>,
+    pub deadlines: Deadlines<BN>,
+}
+
+#[derive(Clone, Decode, Encode)]
+pub enum LegacyMarketDisputeMechanism<AI> {
+    Authorized(AI),
+    Court,
+    SimpleDisputes,
+}
+
+type MomentOf<T> = <<T as Config>::Timestamp as frame_support::traits::Time>::Moment;
+
+type LegacyMarketOf<T> = LegacyMarket<
+    <T as frame_system::Config>::AccountId,
+    <T as frame_system::Config>::BlockNumber,
+    MomentOf<T>,
+>;
+type MarketOf<T> = Market<
+    <T as frame_system::Config>::AccountId,
+    <T as frame_system::Config>::BlockNumber,
+    MomentOf<T>,
+>;
+
+pub struct UpdateMarketsForAuthorizedMDM<T>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for UpdateMarketsForAuthorizedMDM<T> {
+    fn on_runtime_upgrade() -> frame_support::weights::Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(1);
+        let storage_version = utility::get_on_chain_storage_version_of_market_commons_pallet();
+        if storage_version != MARKET_COMMONS_REQUIRED_STORAGE_VERSION {
+            log::info!("Skipping UpdateMarketsForAuthorizedMDM; markets already up to date");
+            return total_weight;
+        }
+        let mut new_markets_data: Vec<(Vec<u8>, MarketOf<T>)> = Vec::new();
+        for (key, legacy_market) in storage_iter::<LegacyMarketOf<T>>(MARKET_COMMONS, MARKETS) {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().reads(1));
+            let dispute_mechanism = match legacy_market.dispute_mechanism {
+                LegacyMarketDisputeMechanism::Authorized(_) => MarketDisputeMechanism::Authorized,
+                LegacyMarketDisputeMechanism::Court => MarketDisputeMechanism::Court,
+                LegacyMarketDisputeMechanism::SimpleDisputes => {
+                    MarketDisputeMechanism::SimpleDisputes
+                }
+            };
+            let new_market = Market {
+                creator: legacy_market.creator,
+                creation: legacy_market.creation,
+                creator_fee: legacy_market.creator_fee,
+                oracle: legacy_market.oracle,
+                metadata: legacy_market.metadata,
+                market_type: legacy_market.market_type,
+                period: legacy_market.period,
+                scoring_rule: legacy_market.scoring_rule,
+                status: legacy_market.status,
+                report: legacy_market.report,
+                resolved_outcome: legacy_market.resolved_outcome,
+                dispute_mechanism,
+                deadlines: legacy_market.deadlines,
```

Comment:
For compiler it can be hard to implement or perhaps algorithm may be very time consuming.

Reply:
It doesn't work, because the new market has pedantically a different type than the old market.

Reply:
> For compiler it can be hard to implement or perhaps algorithm may be very time consuming.

I guess. Still, would be cool if we could do this. ;)

> It doesn't work, because the new market has pedantically a different type than the old market.

Yeah, but the remaining fields all have the same type. That's why I had hoped this would work.